buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/java/AvailableJavaInstallations.kt

```diff
@@ -66,71 +62,73 @@ private
-    private
-    val javaInstallations: Map<JavaVersion, JavaInstallation>
 
     val currentJavaInstallation: JavaInstallation
     val javaInstallationForTest: JavaInstallation
+    val javaInstallationForCompilation: JavaInstallation
 
     init {
-        val resolvedJava7Home = resolveJavaHomePath(java7HomePropertyName, project)
-        val resolvedJava9Home = resolveJavaHomePath(java9HomePropertyName, project)
-        require(resolvedJava9Home != null || JavaVersion.current().isJava9Compatible) { "Building gradle on Java 7/8 requires $java9HomePropertyName system property or project property" }
-        val javaHomesForCompilation = listOfNotNull(resolvedJava7Home, resolvedJava9Home)
-        val javaHomeForTest = resolveJavaHomePath(testJavaHomePropertyName, project)
-        javaInstallations = findJavaInstallations(javaHomesForCompilation)
         currentJavaInstallation = JavaInstallation(true, Jvm.current(), JavaVersion.current(), javaInstallationProbe)
-        javaInstallationForTest = when (javaHomeForTest) {
-            null -> currentJavaInstallation
-            else -> detectJavaInstallation(javaHomeForTest)
-        }
+        javaInstallationForTest = determineJavaInstallation(testJavaHomePropertyName)
+        javaInstallationForCompilation = determineJavaInstallation(compileJavaHomePropertyName)
     }
 
-    fun jdkForCompilation(javaVersion: JavaVersion) = when {
-        javaVersion == currentJavaInstallation.javaVersion -> currentJavaInstallation
-        javaInstallations.containsKey(javaVersion) -> javaInstallations[javaVersion]!!
-        javaVersion <= currentJavaInstallation.javaVersion -> currentJavaInstallation
-        else -> throw IllegalArgumentException("No Java installation found which supports Java version $javaVersion")
+    private
+    fun determineJavaInstallation(propertyName: String): JavaInstallation {
+        val resolvedJavaHome = resolveJavaHomePath(propertyName)
+        when (resolvedJavaHome) {
+            null -> return currentJavaInstallation
+            else -> return detectJavaInstallation(resolvedJavaHome)
+        }
     }
 
-    fun validateForRemoteBuildCacheUsage() {
-        val validationErrors = validateCompilationJdks()
-        if (validationErrors.isNotEmpty()) {
-            val message = formatValidationError(
-                "In order to have cache hits from the remote build cache, your environment needs to be configured accordingly!",
-                validationErrors
-            )
-            throw GradleException(message)
+    fun validateForAllBuilds() {
+        if (project.gradle.startParameter.isBuildCacheEnabled && currentJavaInstallation.vendorAndMajorVersion != oracleJdk8) {
+            System.err.println("To leverage remote cache, you'd better use Oracle JDK 8 to perform this build. It's currently Is currently ${currentJavaInstallation.vendorAndMajorVersion} at ${currentJavaInstallation.javaHome}")
```

Comment:
We should have the validation for possible cache hits cause an error. It can be part of the old validation task I guess.

For now, running the build with Oracle Java 8 or Java 9 should be fine. Compilation should be done by Oracle JDK 9.